### DISCUSSION TITLE
API rule duplicates numeric value in @request.data when doing comparison with a view table.

### DISCUSSION THREAD

**Discussion by fisocodes** [pocketbase/pocketbase#5165] [Q&A]
I'm trying to define an API rule for an inventory.

For creating a product exit, the quantity in exit should be less or equal to the product stock.

Tables are defined as followed.

For productStats (view collection):
![image](https://github.com/pocketbase/pocketbase/assets/66647293/b51c26ca-ba96-4ccb-b805-890f3171b3da)

For exits (base collection):
![image](https://github.com/pocketbase/pocketbase/assets/66647293/318913bc-a620-45c7-be91-93d1ffaaf515)

The create rule for exits is the following:
![image](https://github.com/pocketbase/pocketbase/assets/66647293/396e4832-3bc4-4e47-883a-6c805bc5a0dc)

This API rule fails when:
@request.data.quantity > (productId.productStats_via_productId.stock / 2)

For example if stock equals 5, the rule fails when the exit quantity is greater than 2.5, However, when exit quantity equals or is less than 2.5 the API rule works. However when the API rule pass, the exit stores the right value in quantity not the double.

So, mi guess the API rule is duplicating the exit quantity during comparison.

Has anyone been into a similar issue?

**Comment by ganigeorgiev** [comment]
~What is happening is probably the query resulting in a number vs json string comparison (eg. `5 < '2.5'`).~ _On second read if the `stock` value depends on the created record's collection, then your issue is most likely caused by the dry submit - see the reply below._

What you could try is to explicitly cast your View collection columns to `REAL` value so that they can be mapped to the correct `number` field, aka. change your view query SELECT fields to something like:

```sql
SELECT
   ...
   CAST(stock as REAL) as stock,
   ...etc.
```

**Reply by ganigeorgiev** [reply]
Also keep in mind that the Create API rule currently executes **after** a dry submit of the record, which could affect your view depending on how is constructed.

**Reply by fisocodes** [reply]
> ~What is happening is probably the query resulting in a number vs json string comparison (eg. `5 < '2.5'`).~ _On second read if the `stock` value depends on the created record's collection, then your issue is most likely caused by the dry submit - see the reply below._
> 
> What you could try is to explicitly cast your View collection columns to `REAL` value so that they can be mapped to the correct `number` field, aka. change your view query SELECT fields to something like:
> 
> ```sql
> SELECT
>    ...
>    CAST(stock as REAL) as stock,
>    ...etc.
> ```

I have the same issue when using integers, for example product stock = 5 and exit quantity = 3.

And yes, I get the following error:

> DrySubmit create rule failure: sql: no rows in result set

Which, tbh I have no idea what does that mean. I'll try to research about it, I would appreciate any guidance, thank you!

**Reply by ganigeorgiev** [reply]
What I meant by _"Create API rule currently executes after a dry submit of the record"_ is that the Create API rule is checked **after** the record is inserted in the database (it is persisted only if the rule succeed).

Because views are computed, if `productStats` depends on the `exits` collection then it will include also the just inserted record in its aggregation. If that's the case, then there is no workaround for this at the moment (_task to consider changing how the Create Rule is checked without dry submit exists in my todo but remains a low priority for now_).

If you need further help, please export your collections configuration as json (_Admin UI > Settings > Export collections_).

**Reply by fisocodes** [reply]
Right, understood. That makes sense. The only work around I think of right now is to validate my data on the front end. Thank you very much for your time.

**Comment by fisocodes** [comment]
For now I will focus on validate my data on the front end. Dry submit will not allow me to validate my data correctly in the API rules.
Another work around could be validating data with hooks, adding a little more complexity to the implementation.

It could be a great idea to add the "Dry submit" behaviour in the API rules documentation as a note or warning.

**Reply by ganigeorgiev** [reply]
_Just in case someone stumble here: the dry submit behavior is discussed to change in v0.24+ per https://github.com/pocketbase/pocketbase/discussions/6073._
